File: kotlin_obdii/app/src/main/java/com/rheosoft/obdii/viewmodels/GaugesViewModel.kt
package com.rheosoft.obdii.viewmodels

import com.rheosoft.obdii.core.ConfigData
import com.rheosoft.obdii.core.GaugesDisplayMode
import com.rheosoft.obdii.core.PIDStats
import com.rheosoft.obdii.core.PidInterestRegistry
import com.rheosoft.obdii.core.PidListProviding
import com.rheosoft.obdii.core.PidStatsProviding
import com.rheosoft.obdii.core.UnitsProviding
import com.rheosoft.obdii.core.DefaultPidStore
import com.rheosoft.obdii.core.PidStore
import com.rheosoft.obdii.core.OBDConnectionManager
import com.rheosoft.obdii.models.ObdPidKind
import com.rheosoft.obdii.models.ObdiiPid
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.launch

data class GaugeTile(val id: String, val pid: ObdiiPid, val stats: PIDStats?)
data class GaugesUiState(val tiles: List<GaugeTile>, val displayMode: GaugesDisplayMode)

class GaugesViewModel(
    private val pidProvider: PidListProviding = DefaultPidStore,
    private val statsProvider: PidStatsProviding = OBDConnectionManager,
    private val unitsProvider: UnitsProviding = ConfigData,
    private val interestRegistry: PidInterestRegistry = PidInterestRegistry.instance,
    private val scope: CoroutineScope = CoroutineScope(Dispatchers.Default + Job()),
) : BaseViewModel() {
    private val interestToken = interestRegistry.makeToken()
    private var isVisible = false
    var tiles: List<GaugeTile> = emptyList()
        private set
    private val _uiStateStream = MutableStateFlow(GaugesUiState(tiles, unitsProvider.gaugesDisplayMode))
    val uiStateStream: StateFlow<GaugesUiState> = _uiStateStream.asStateFlow()

    val isEmpty: Boolean
        get() = tiles.isEmpty()

    init {
        scope.launch { pidProvider.pidsStream.collectLatest { rebuildTiles() } }
        scope.launch { statsProvider.pidStatsStream.collectLatest { rebuildTiles() } }
        scope.launch { unitsProvider.unitsStream.collectLatest { rebuildTiles() } }
        scope.launch { unitsProvider.gaugesDisplayModeStream.collectLatest { rebuildTiles() } }
        rebuildTiles()
    }

    fun setVisible(visible: Boolean) {
        if (isVisible == visible) return
        isVisible = visible
        updateInterest()
    }

    fun setDisplayMode(mode: GaugesDisplayMode) {
        unitsProvider.gaugesDisplayMode = mode
    }

    suspend fun moveEnabled(from: Int, to: Int) {
        val store = pidProvider as? PidStore ?: return
        store.moveEnabled(from, to)
    }

    private fun rebuildTiles() {
        val newTiles = pidProvider.pids
            .filter { it.enabled && it.kind == ObdPidKind.gauge }
            .map { GaugeTile(it.id, it, statsProvider.statsFor(it.pidCommand)) }
        val mode = unitsProvider.gaugesDisplayMode
        if (newTiles == tiles && mode == _uiStateStream.value.displayMode) return
        tiles = newTiles
        _uiStateStream.value = GaugesUiState(tiles, mode)
        updateInterest()
        notifyChanged()
    }

    private fun updateInterest() {
        if (!isVisible) {
            scope.launch { interestRegistry.clear(interestToken) }
            return
        }
        interestRegistry.replace(tiles.map { it.pid.pidCommand }.toSet(), interestToken)
    }
}

File: flutter_obdii/lib/viewmodels/gauges_viewmodel.dart
// Port of GaugesViewModel.swift — Jim Mittler
// ViewModel for the selected gauges dashboard.
// Keeps an array of visible tiles. Updates when new data arrives.
// Combines PidListProviding + PidStatsProviding + UnitsProviding.

import 'dart:async';

import '../core/config_data.dart';
import '../core/obd_connection_manager.dart';
import '../core/pid_interest_registry.dart';
import '../core/pid_store.dart';
import '../core/obdiipid.dart';
import 'base_view_model.dart';

// ─────────────────────────────────────────────
// MARK: Tile (replaces Swift Tile struct)
// ─────────────────────────────────────────────

class GaugeTile {
  final String id;
  final ObdiiPid pid;
  final PIDStats? stats;

  const GaugeTile({required this.id, required this.pid, this.stats});

  @override
  bool operator ==(Object other) =>
      other is GaugeTile &&
      other.id == id &&
      other.stats?.sampleCount == stats?.sampleCount &&
      other.stats?.latest.value == stats?.latest.value;

  @override
  int get hashCode => Object.hash(id, stats?.sampleCount);
}

// ─────────────────────────────────────────────
// MARK: GaugesViewModel
// ─────────────────────────────────────────────

class GaugesViewModel extends BaseViewModel {
  final PidListProviding _pidProvider;
  final PidStatsProviding _statsProvider;
  final UnitsProviding _unitsProvider;
  final PidInterestRegistry _interestRegistry;
  final String _interestToken;

  List<GaugeTile> _tiles = [];
  List<GaugeTile> get tiles => _tiles;
  bool _isVisible = false;

  bool get isEmpty => _tiles.isEmpty;

  StreamSubscription? _pidSub;
  StreamSubscription? _statsSub;
  StreamSubscription? _unitsSub;

  GaugesViewModel({
    PidListProviding? pidProvider,
    PidStatsProviding? statsProvider,
    UnitsProviding? unitsProvider,
    PidInterestRegistry? interestRegistry,
  })  : _pidProvider = pidProvider ?? PidStore.instance,
        _statsProvider = statsProvider ?? OBDConnectionManager.instance,
        _unitsProvider = unitsProvider ?? ConfigData.instance,
        _interestRegistry = interestRegistry ?? PidInterestRegistry.instance,
        _interestToken =
            (interestRegistry ?? PidInterestRegistry.instance).makeToken() {
    _bind();
  }

  void _bind() {
    // Rebuild whenever pids list changes
    _pidSub = _pidProvider.pidsStream.listen((_) => _rebuildTiles());

    // Rebuild whenever stats update
    _statsSub = _statsProvider.pidStatsStream.listen((_) => _rebuildTiles());

    // Rebuild whenever units change (values need re-conversion)
    _unitsSub = _unitsProvider.unitsStream.listen((_) => _rebuildTiles());

    // Initial build
    _rebuildTiles();
  }

  void _rebuildTiles() {
    final newTiles = _pidProvider.pids
        .where((p) => p.enabled && p.kind == ObdPidKind.gauge)
        .map((pid) => GaugeTile(
              id: pid.id,
              pid: pid,
              stats: _statsProvider.statsFor(pid.pidCommand),
            ))
        .toList();

    // Avoid unnecessary notify if tiles are the same
    if (_tilesEqual(newTiles, _tiles)) return;
    _tiles = newTiles;
    _updateInterest();
    notifyListeners();
  }

  void setVisible(bool isVisible) {
    if (_isVisible == isVisible) return;
    _isVisible = isVisible;
    _updateInterest();
  }

  void _updateInterest() {
    if (!_isVisible) {
      _interestRegistry.clear(_interestToken);
      return;
    }
    final interested = _tiles.map((t) => t.pid.pidCommand).toSet();
    _interestRegistry.replace(interested, _interestToken);
  }

  bool _tilesEqual(List<GaugeTile> a, List<GaugeTile> b) {
    if (a.length != b.length) return false;
    for (var i = 0; i < a.length; i++) {
      if (a[i] != b[i]) return false;
    }
    return true;
  }

  @override
  void dispose() {
    _pidSub?.cancel();
    _statsSub?.cancel();
    _unitsSub?.cancel();
    _interestRegistry.clear(_interestToken);
    super.dispose();
  }
}
